x/programs/program/memory.go

```diff
@@ -164,31 +167,52 @@ type CallParam struct {
 
 // WriteParams is a helper function that writes the given params to memory if non integer.
 // Supported types include int, uint64 and string.
-func WriteParams(m *Memory, p []CallParam) ([]uint64, error) {
-	params := []uint64{}
+func WriteParams(m *Memory, p []CallParam) ([]SmartPtr, error) {
+	var params []SmartPtr
 	for _, param := range p {
 		switch v := param.Value.(type) {
-		case string:
-			ptr, err := WriteBytes(m, []byte(v))
+		case ids.ID:
+			smartPtr, err := BytesToSmartPtr(v[:], m)
 			if err != nil {
 				return nil, err
 			}
-			params = append(params, uint64(ptr))
-		case int:
-			if v < 0 {
-				return nil, fmt.Errorf("failed to write param: %w", ErrUnderflow)
+			params = append(params, smartPtr)
+		case string:
+			smartPtr, err := BytesToSmartPtr([]byte(v), m)
+			if err != nil {
+				return nil, err
 			}
-			params = append(params, uint64(v))
-		case uint64:
+			params = append(params, smartPtr)
+		case SmartPtr:
 			params = append(params, v)
 		default:
-			return nil, fmt.Errorf("invalid param: supported types int, uint64 and string")
+			ptr, err := argumentToSmartPtr(v, m)
+			if err != nil {
+				return nil, err
+			}
+			params = append(params, ptr)
 		}
 	}
 
 	return params, nil
 }
 
+// SerializeParameter serializes [obj] using Borsh
+func serializeParameter(obj interface{}) ([]byte, error) {
+	bytes, err := borsh.Serialize(obj)
```

Comment:
I think the serialization should be done outside of `memory.go`. That way we can abstract away the serializer(borsh) and just require all params to just be bytes/SmartPtrs.

Reply:
I'll make this same change when I redo the write params.  I'd like to move all of that logic out, but I'd like to do it in the followup.

Reply:
for now, moving this out of the program module and into simulator, since it isn't used anywhere else.